### PR TITLE
Visibility

### PR DESCRIPTION
**Pull request by insertinterestingnamehere**
These are the minimal changes necessary to get everything working on Windows. A lot of refactoring still needs to be done, but it was best to just get things working again first.

### CODE REVIEW

**Comment by insertinterestingnamehere** [comment]
Hmm. I'm not sure why Appveyor complains and my local machine does not. It looks like it's finding that various symbols are multiply defined (in both libdynd and libpydynd). The two libraries appear to be using all the same compiler settings though, so they should be linking against the same standard library, and all of this should just work. I could work around this by using the flag `/FORCE:MULTIPLE` but that is something I'd rather avoid.

**Comment by insertinterestingnamehere** [comment]
@mwiebe Does this fix everything for you? Any thoughts on what's going on with appveyor?

**Comment by insertinterestingnamehere** [comment]
Hmm. It looks like when libdynd is built inside dynd-python it defaults to building a static library. That could be the source of the errors. I'm investigating now.

File: dynd/__init__.py [hunk 1]
```diff
@@ -1,3 +1,21 @@
...
+    import sys
+    from ctypes import cdll
+    is_64_bit = sys.maxsize > 2**32
+    if not os.path.isfile("libdynd.dll"):
+        if is_64_bit:
+            if os.path.isfile("c:/Program Files/libdynd/lib/libdynd.dll"):
+                cdll.LoadLibrary("c:/Program Files/libdynd/lib/libdynd.dll")
+        else:
+            if os.path.isfile("c:/Program Files (x86)/libdynd/lib/libdynd.dll"):
+                cdll.LoadLibrary("c:/Program Files (x86)/libdynd/lib/libdynd.dll")
```
Comment (mwiebe) [inline dynd/__init__.py]: I think this logic would be better using os.environ['ProgramFiles'] and/or os.environ['ProgramFiles(x86)'].

**Comment by mwiebe** [comment]
@insertinterestingnamehere after rebasing against master, it gets pretty close to building. Like this:

```
C:\Dev\dynd-python\build\temp.win-amd64-3.4\Release\type.cxx(10230): error C2220: warning treated as error - no 'object' file gene
rated [C:\Dev\dynd-python\build\temp.win-amd64-3.4\Release\type.vcxproj]
```

Lots of additional compiler warnings from the cython-generated code.

**Comment by insertinterestingnamehere** [comment]
@mwiebe Thanks for trying it out. I'll update the logic searching for libdynd, rebase and then see if I can replicate the warning-as-error issue.

**Comment by mwiebe** [comment]
I'm building with `python setup.py bdist_wheel`, same as the appveyor.yml uses. Is that how you're building it presently?

**Comment by insertinterestingnamehere** [comment]
Currently I'm using `python setup.py install`.

**Comment by insertinterestingnamehere** [comment]
Are you building libdynd as a subproject as well?

**Comment by mwiebe** [comment]
Yes, I've got libdynd in the dynd_python/libraries dir.

**Comment by insertinterestingnamehere** [comment]
Got it. I'll try that setup.

**Comment by mwiebe** [comment]
After removing "/WX" (warnings as errors), get a lot of linker errors, ending with:

```
  libdynd.lib(base_tuple_type.obj) : error LNK2005: "public: virtual void __cdecl dynd::ndt::base_tuple_type::foreach_leading(char const *,char *,void (__cdecl*)(class dynd::ndt::
type const &,char const *,char *,void *),void *)const " (?foreach_leading@base_tuple_type@ndt@dynd@@UEBAXPEBDPEADP6AXAEBVtype@23@01PEAX@Z3@Z) already defined in pydynd.lib(pydynd.
dll) [C:\Dev\dynd-python\build\temp.win-amd64-3.4\Release\type.vcxproj]
  libdynd.lib(base_tuple_type.obj) : error LNK2005: "public: virtual unsigned __int64 __cdecl dynd::ndt::base_tuple_type::get_default_data_size(void)const " (?get_default_data_siz
e@base_tuple_type@ndt@dynd@@UEBA_KXZ) already defined in pydynd.lib(pydynd.dll) [C:\Dev\dynd-python\build\temp.win-amd64-3.4\Release\type.vcxproj]
  libdynd.lib(base_tuple_type.obj) : error LNK2005: "public: virtual void __cdecl dynd::ndt::base_tuple_type::get_shape(__int64,__int64,__int64 *,char const *,char const *)const "
 (?get_shape@base_tuple_type@ndt@dynd@@UEBAX_J0PEA_JPEBD2@Z) already defined in pydynd.lib(pydynd.dll) [C:\Dev\dynd-python\build\temp.win-amd64-3.4\Release\type.vcxproj]
  libdynd.lib(base_tuple_type.obj) : error LNK2005: "public: virtual void __cdecl dynd::ndt::base_tuple_type::get_vars(class std::unordered_set<class std::basic_string<char,struct
 std::char_traits<char>,class std::allocator<char> >,struct std::hash<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >,struct std::equal_t
o<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >,class std::allocator<class std::basic_string<char,struct std::char_traits<char>,class s
td::allocator<char> > > > &)const " (?get_vars@base_tuple_type@ndt@dynd@@UEBAXAEAV?$unordered_set@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@U?$hash@V?$basic_st
ring@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@2@U?$equal_to@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@2@V?$allocator@V?$basic_string@DU?$char_traits@D@s
td@@V?$allocator@D@2@@std@@@2@@std@@@Z) already defined in pydynd.lib(pydynd.dll) [C:\Dev\dynd-python\build\temp.win-amd64-3.4\Release\type.vcxproj]
  libdynd.lib(base_tuple_type.obj) : error LNK2005: "public: virtual bool __cdecl dynd::ndt::base_tuple_type::is_expression(void)const " (?is_expression@base_tuple_type@ndt@dynd@@
UEBA_NXZ) already defined in pydynd.lib(pydynd.dll) [C:\Dev\dynd-python\build\temp.win-amd64-3.4\Release\type.vcxproj]
  libdynd.lib(base_tuple_type.obj) : error LNK2005: "public: virtual bool __cdecl dynd::ndt::base_tuple_type::is_unique_data_owner(char const *)const " (?is_unique_data_owner@base
_tuple_type@ndt@dynd@@UEBA_NPEBD@Z) already defined in pydynd.lib(pydynd.dll) [C:\Dev\dynd-python\build\temp.win-amd64-3.4\Release\type.vcxproj]
  libdynd.lib(base_tuple_type.obj) : error LNK2005: "public: virtual bool __cdecl dynd::ndt::base_tuple_type::match(char const *,class dynd::ndt::type const &,char const *,class s
td::map<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class dynd::ndt::type,struct std::less<class std::basic_string<char,struct std::cha
r_traits<char>,class std::allocator<char> > >,class std::allocator<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const ,
class dynd::ndt::type> > > &)const " (?match@base_tuple_type@ndt@dynd@@UEBA_NPEBDAEBVtype@23@0AEAV?$map@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@Vtype@ndt@dyn
d@@U?$less@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@2@V?$allocator@U?$pair@$$CBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@Vtype@ndt@dynd@@
@std@@@2@@std@@@Z) already defined in pydynd.lib(pydynd.dll) [C:\Dev\dynd-python\build\temp.win-amd64-3.4\Release\type.vcxproj]
  libdynd.lib(base_tuple_type.obj) : error LNK2005: "public: virtual void __cdecl dynd::ndt::base_tuple_type::print_data(class std::basic_ostream<char,struct std::char_traits<char
> > &,char const *,char const *)const " (?print_data@base_tuple_type@ndt@dynd@@UEBAXAEAV?$basic_ostream@DU?$char_traits@D@std@@@std@@PEBD1@Z) already defined in pydynd.lib(pydynd.
dll) [C:\Dev\dynd-python\build\temp.win-amd64-3.4\Release\type.vcxproj]
  libdynd.lib(fixed_dim_kind_type.obj) : error LNK2005: "public: __cdecl dynd::ndt::fixed_dim_kind_type::fixed_dim_kind_type(class dynd::ndt::type const &)" (??0fixed_dim_kind_typ
e@ndt@dynd@@QEAA@AEBVtype@12@@Z) already defined in pydynd.lib(pydynd.dll) [C:\Dev\dynd-python\build\temp.win-amd64-3.4\Release\type.vcxproj]
  libdynd.lib(type_type.obj) : error LNK2005: "public: __cdecl dynd::ndt::type_type::type_type(void)" (??0type_type@ndt@dynd@@QEAA@XZ) already defined in pydynd.lib(pydynd.dll) [C
:\Dev\dynd-python\build\temp.win-amd64-3.4\Release\type.vcxproj]
  C:\Dev\dynd-python\build\temp.win-amd64-3.4\Release\ndt\Release\type.pyd : fatal error LNK1169: one or more multiply defined symbols found [C:\Dev\dynd-python\build\temp.win-amd
64-3.4\Release\type.vcxproj]

    62 Warning(s)
    110 Error(s)
```

**Comment by insertinterestingnamehere** [comment]
Yep. That's the error that's showing up on appveyor as well. My theory is that those stem from linking libdynd statically and libpydynd dynamically. When I use dlls all around, the issue goes away. I haven't found why libdynd is still compiling as a static library when it's in dynd-python's source tree though.

**Comment by insertinterestingnamehere** [comment]
Okay, I just managed to reproduce this locally.

**Comment by insertinterestingnamehere** [comment]
Success! Windows builds are now working. There's plenty of refactoring still needed in the long run, but this should be good to go.

**Comment by izaid** [comment]
Yes!!! Really great work!

**Comment by mwiebe** [comment]
@insertinterestingnamehere Works for me locally too, awesome!

**Pull request merged**